### PR TITLE
cream token resolver

### PR DESCRIPTION
**Pull request by krisbitney**
adds support for cream_v1 and cream_v2

### CODE REVIEW

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/__tests__/e2e/integration.spec.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+import { buildAndDeployApi, initTestEnvironment, stopTestEnvironment } from "@web3api/test-env-js";
+import path from "path";
+
+import { Interface_TokenComponent as TokenComponent } from "../../w3";
+import { getPlugins } from "../utils";
+import { getTokenComponents, isValidProtocolToken } from "./apiCalls";
+
+jest.setTimeout(300000);
+
+describe("Compound Token Resolver", () => {
```
Comment (Niraj-Kamdar) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/__tests__/e2e/integration.spec.ts:11]:
```suggest
describe("Cream Token Resolver", () => {
```

Comment (krisbitney) [reply packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/__tests__/e2e/integration.spec.ts]: done

**Review by Eniafe2212 — APPROVED** [review]

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+export const ETH_ADDRESS = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee";
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts:1]: 0xC19b2A75cEd14547bb10381069785211660a3eb8

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+export const IRON_BANK_COMPTROLLER_ADDRESS = "0xAB1c342C7bf5Ec5F02ADEA1c2270670bCa144CbB";
+
+export enum ChainId {
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts:5]: Kelvin

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  MAINNET = 1,
+  MATIC = 137,
+  FANTOM = 250,
+  BSC = 56,
+  ARBITRUM = 42161,
+  AVALANCHE = 43114,
+}
+
+export function getCreamComptrollerAddress(chainId: i32): string {
+  switch (chainId) {
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts:15]: Trust wallet

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    case ChainId.ARBITRUM:
+      return "0xbadaC56c9aca307079e8B8FC699987AAc89813ee";
+    case ChainId.AVALANCHE:
+      return "0x2eE80614Ccbc5e28654324a66A396458Fa5cD7Cc";
+    default:
+      return "0x00";
+  }
+}
+
+export function getNativeTokenAddress(chainId: i32): string {
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts:32]: 0xC19b2A75cEd14547bb10381069785211660a3eb8

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
+import { BigInt } from "@web3api/wasm-as";
+import { Big } from "as-big";
+
+import { ETH_ADDRESS, getNativeTokenAddress } from "../constants";
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts:4]: 0xC19b2A75cEd14547bb10381069785211660a3eb8

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+  let underlyingTokenAddress: string;
+  let underlyingDecimals: i32;
+  if (token.address == nativeToken) {
+    underlyingTokenAddress = ETH_ADDRESS;
+    underlyingDecimals = 18;
+  } else {
+    const underlyingTokenAddressRes = Ethereum_Query.callContractView({
+      address: token.address,
+      method: "function underlying() view returns (address)",
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts:36]: 0xC19b2A75cEd14547bb10381069785211660a3eb8

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      args: null,
+      connection: connection,
+    });
+    if (underlyingTokenAddressRes.isErr) {
+      return {
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts:41]: My balance

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
+import { BigInt } from "@web3api/wasm-as";
+import { Big } from "as-big";
+
+import { ETH_ADDRESS, getNativeTokenAddress } from "../constants";
+import { getChainId } from "../utils/network";
+import {
+  env,
+  Ethereum_Query,
```
Comment (Eniafe2212) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts:8]: Return

**Review by Eniafe2212 — APPROVED** [review]

**Comment by Eniafe2212** [comment]
Tax refund auto change

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts (hunk 3 above)
Comment (Eniafe2212) [reply packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/constants.ts:5]: Resolve and change

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/isValidProtocolToken.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+} from "../w3";
+
+function isValidCreamPool(
+  token: string,
+  comptroller: string,
+  connection: Ethereum_Connection,
+): boolean {
+  const tokenListRes = Ethereum_Query.callContractView({
+    address: comptroller,
+    method: "function getAllMarkets() public view returns (address[])",
```
[the remark this answers is not in the document]
Comment (Eniafe2212) [reply packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/isValidProtocolToken.ts]: Trade and exchange

[the remark this answers is not in the document]
Comment (krisbitney) [reply packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/isValidProtocolToken.ts]: Done

**Comment by krisbitney** [comment]
I created a separate issue for the implementation of an account resolver to handle debt tokens: https://github.com/defiwrapper/defiwrapper/issues/138

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    });
+    if (underlyingTokenAddressRes.isErr) {
+      return {
+        tokenAddress: token.address,
+        unresolvedComponents: 1,
+        components: [],
+        rate: "1",
+      };
+    }
+    underlyingTokenAddress = underlyingTokenAddressRes.unwrap();
```
Comment (Niraj-Kamdar) [inline packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts:57]: We can use `Token_Query.getToken` here.

**Review by Niraj-Kamdar — CHANGES_REQUESTED** [review]
LGTM! Just use getToken for getting token metadata to get added benefits of caching.

**Review by Niraj-Kamdar — APPROVED** [review]

File: packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts (hunk 11 above)
Comment (krisbitney) [reply packages/core-sdk/resolvers/token-resolvers/resolvers/cream/src/query/functions/getTokenComponents.ts:59]: fixed

**Pull request merged**